core/embed/bootloader/version_check.c

```diff
@@ -0,0 +1,52 @@
+/*
+ * This file is part of the Trezor project, https://trezor.io/
+ *
+ * Copyright (c) SatoshiLabs
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <string.h>
+
+#include "version_check.h"
+
+#include "error_handling.h"
+#include "model_version.h"
+#include "monoctr.h"
+
+void ensure_bootloader_min_version(void) {
+  ensure(
+      monoctr_write(MONOCTR_BOOTLOADER_VERSION, BOOTLOADER_MONOTONIC_VERSION),
+      NULL);
+  uint8_t val = 0;
+  ensure(monoctr_read(MONOCTR_BOOTLOADER_VERSION, &val), NULL);
+  ensure(sectrue * (val == BOOTLOADER_MONOTONIC_VERSION),
+         "Bootloader downgrade protection");
+}
+
+secbool check_firmware_min_version(uint8_t check_version) {
+  uint8_t min_version = 0;
+  ensure(monoctr_read(MONOCTR_FIRMWARE_VERSION, &min_version), "monoctr read");
+
+  return (check_version >= min_version) * sectrue;
+}
+
+void update_firmware_min_version(uint8_t version) {
+  uint8_t min_version = 0;
+  ensure(monoctr_read(MONOCTR_FIRMWARE_VERSION, &min_version), "monoctr read");
+
+  if (version > min_version) {
+    ensure(monoctr_write(MONOCTR_FIRMWARE_VERSION, version), "monoctr write");
+  }
+}
```

Comment:
That would not work properly on U5, where the counter is written in regular flash in secret area (via monoctr, secret and flash_area drivers), this would fail on `ensure` in the secret implementation as it would fail to write the lower version.

The `ensure_bootloader_min_version` is ok in this regard as its used exclusively on F4, where we writee in OTP that does not have these checks. But maybe we should be more explicit in this file that the bootloader function is only meant for F4.

Thus, if anything, i would change the `ensure_bootloader_min_version` to work like the fw one. But then again the original looks a bit more glitch resistant so i would go for keeping it this way.

Reply:
Modified the monoctr so that it returns secfalse in any case it cannot write the new value.

then, i changed the `update_firmware_min_version` to match the `ensure_bootloader_min_version` variant. I did remove the first ensure in that function, as writing would fail immediately now and later we have better error message.

see https://github.com/trezor/trezor-firmware/commit/12298d3c1e3acc9e272fef5006d60a73b4ed92eb and https://github.com/trezor/trezor-firmware/commit/e8b1dceac086565db773391d890f3916863284e4